src/tess.py

```diff
@@ -0,0 +1,364 @@
+
+            # verify that all data for the same oid
+            # has the same filter
+            if len(list(set(df[filt_col]))) > 1:
+                raise ValueError(
+                    f"Expected all data for the same oid to have the same filter, "
+                    f"but the oid {df['oid'].iloc[0]} had filters {filters}."
+                )
+            oid_dfs.append(new_oid_df)
+
+        # split the data into lightcurves
+        # based on the gap between observations
+        data = pd.concat(oid_dfs)
+        data_sort = data.sort_values(by=[time_col], inplace=False)
+        data_sort.reset_index(drop=True)
+
+        dt = np.diff(mjd_conversion(data_sort[time_col]))
+
+        gaps = np.where(dt > gap)[0]
+        gaps = np.append(gaps, len(data))
+        prev_idx = 0
+
+        dfs = []
+        for idx in gaps:
+            df_gap = data_sort[prev_idx : idx + 1]
+            for filt in filters:
+                df = df_gap[df_gap[filt_col] == filt].reset_index(
+                    drop=True, inplace=False
+                )
+
+                # drop the flagged or NaN values
+                if drop_bad:
+                    bad_idx = (df[flag_col] != 0) | (df[mag_col].isna())
+                    df = df[~bad_idx]
+
+                if len(df) > 0:
+                    dfs.append(df)
+
+            prev_idx = idx + 1
+
+        new_datasets = []
+        keep_columns = [
+            time_col,
+            exp_col,
+            filt_col,
+            mag_col,
+            magerr_col,
+            ra_col,
+            dec_col,
+            flag_col,
+        ]
+        for df in dfs:
+            df = df[keep_columns]
+            if len(df) > 0:
+                new_datasets.append(Lightcurve(data=df, **init_kwargs))
+
+        return new_datasets
+
+    def fetch_data_from_observatory(self, cat_row, verbose=0, **kwargs):
+        """
+        Fetch data from TESS for a given source.
+        Must return a dataframe (or equivalent), even if it is an empty one.
+        This must be implemented by each observatory subclass.
+
+        Parameters
+        ----------
+        cat_row: dict like
+            A row in the catalog for a specific source.
+            In general, this row should contain the following keys:
+            name, ra, dec, mag, filter_name (saying which band the mag is in).
+        kwargs: dict
+            Additional keyword arguments to pass to the fetcher.
+
+        Returns
+        -------
+        data : pandas.DataFrame or other data structure
+            Raw data from the observatory, to be put into a RawData object.
+        altdata: dict
+            Additional data to be stored in the RawData object.
+
+        """
+        # TODO: figure out dstArcSec threshold for flagging close stars
+        DST_THRESHOLD = 10
+        # TODO: figure out mag diff threshold for flagging similar stars
+        MAGDIFF_THRESHOLD = 0.5
+        # TODO: figure out optimal radii for queries
```

Comment:
Would I add new parameters by adding add_par() calls in the ParsObsTESS class and setting self._enforce_no_new_attrs=False? Or would the thresholds somehow be passed as part of kwargs for ParsObsTESS, or kwargs for VirtualTESS?

Reply:
You should add these parameters to the `ParsObsTESS` using `add_par`. You put the new parameters in the `__init__` function and at the end put `self._enforce_no_new_attrs = True` to make sure the user doesn't accidentally add parameters that are not defined already. 
Then when you pass a kwarg into the constructor of `VirtualTESS` it will be passed on automatically into `ParsObsTESS` and that would be applied to the parameter values and used in the download function. 
I'm using the following parameter names for ZTF:
`cone_search_radius`, `limiting_magnitude`, `faint_magnitude_difference`, `bright_magnitude_difference`. Let me know if you are not sure what each one is supposed to do.